pkg/api/customization/globalnamespaceaccess/access_common.go

```diff
@@ -6,166 +6,162 @@ import (
+	for key, accessType := range newMemberAccessType {
+		if val, ok := originalMemberAccessType[key]; ok {
+			if val != accessType {
+				requestUpdatesMembers = true
+				break
+			}
+			// mark this original member as present in the current request
+			originalMembersFoundInRequest[key] = true
+		} else {
+			requestUpdatesMembers = true
+			break
 		}
-		projectNS := split[1]
-		prtbs, err := prtbLister.List(projectNS, labels.NewSelector())
-		if err != nil {
-			return err
+	}
+	if requestUpdatesMembers && !ownerAccess {
+		return fmt.Errorf("only members with owner access can add new members")
+	}
+
+	// at this point, all members in the new request have been found in the original global DNS with the same access type
+	// but we need to check if all members from the original global DNS are also present in the current request
+	for member := range originalMemberAccessType {
+		// if any member is not found, this means the current request is updating members list by deleting this member
+		if !originalMembersFoundInRequest[member] && !ownerAccess {
+			return fmt.Errorf("only members with owner access can delete existing members")
 		}
-		prtbMap := make(map[string]string)
-		for _, prtb := range prtbs {
-			if prtb.GroupPrincipalName != "" {
-				prtbMap[prtb.GroupPrincipalName] = prtb.RoleTemplateName
+	}
+	return nil
+}
+
+func (ma *MemberAccess) CheckAccessTypesAllowsUpdates(callerID, creatorID, name string, members []v3.Member) (string, error) {
+	var accessType, username string
+	if callerID == creatorID {
+		return OwnerAccess, nil
+	}
+	for _, m := range members {
+		username = ""
+		if m.UserName == "" && m.UserPrincipalName != "" {
+			user, err := ma.getUserFromUserPrincipalID(m.UserPrincipalName)
+			if err != nil {
+				return "", err
+			}
+			if user == nil {
+				return "", fmt.Errorf("no user found for principal %v", m.UserPrincipalName)
 			}
+			if user.Name == callerID {
+				username = user.Name
+			}
+		} else if m.UserName == callerID {
+			username = m.UserName
 		}
-		// see if all groups are in prtbMap
-		for g, accessType := range groups {
-			if rtName, ok := prtbMap[g]; ok {
-				if resourceType == client.MultiClusterAppType {
-					rt, err := rtLister.Get("", rtName)
-					if err != nil {
-						return err
-					}
-					for _, r := range rt.Rules {
-						if !slice.ContainsString(r.Resources, resource) {
-							continue
-						}
-						if slice.ContainsString(r.Verbs, "*") {
-							continue
-						}
-						for _, v := range accessVerbs[accessType] {
-							if !slice.ContainsString(r.Verbs, v) {
-								return fmt.Errorf("member (group) %v cannot %v %v in project %v", g, v, resource, t)
-							}
-						}
-
-					}
-				}
-			} else {
-				return fmt.Errorf("member (group) %v does not have access to project %v", g, t)
+		if username != "" { // found the caller
+			if m.AccessType == ReadonlyAccess {
+				return "", fmt.Errorf("user %v is read-only member of %v, cannot edit it", callerID, name)
```

Comment:
This check is applicable for multiclusterapp which has 3 roles. But for globalDNS we need to check that the member has accessType=owner?

Reply:
I'm checking that when the method returns, but I can modify this method to accept the set of allowed roles

Reply:
yeah the name of the method is misleading, either change the method name to reflect what it does or do the check here